pkg/daemon/pinned_image_set.go

```diff
@@ -0,0 +1,1273 @@
+			if ok {
+				if imageInfo.Pulled {
+					continue
+				}
+			}
+		}
+
+		exists, err := p.criClient.ImageStatus(ctx, imageName)
+		if err != nil {
+			return err
+		}
+		if exists {
+			// dont account for storage if the image already exists
+			continue
+		}
+
+		// check if the image is already in the cache before fetching the size
+		if value, found := p.cache.Get(imageName); found {
+			imageInfo, ok := value.(imageInfo)
+			if ok {
+				requiredStorage += imageInfo.Size * 2
+				continue
+			}
+			// cache is corrupted, delete the key
+			klog.Warningf("corrupted cache entry for image %q, deleting", imageName)
+			p.cache.Remove(imageName)
+		}
+		size, err := getImageSize(ctx, imageName, p.authFilePath)
+		if err != nil {
+			return err
+		}
+
+		// cache miss
+		p.cache.Add(imageName, imageInfo{Name: imageName, Size: size})
+
+		// account for decompression
+		requiredStorage += size * 2
+	}
+
+	if requiredStorage >= capacity-p.minStorageAvailableBytes.Value() {
+		klog.Errorf("%v capacity=%d, required=%d", errInsufficientStorage, capacity, requiredStorage+p.minStorageAvailableBytes.Value())
+		return errInsufficientStorage
+	}
+
+	return nil
+}
+
+func isImageSetInPool(imageSet string, pool *mcfgv1.MachineConfigPool) bool {
+	for _, set := range pool.Spec.PinnedImageSets {
+		if set.Name == imageSet {
+			return true
+		}
+	}
+	return false
+}
+
+// ensureCrioPinnedImagesConfigFile creates the crio config file which populates pinned_images with a unique list of images then reloads crio.
+// If the list is empty, and a config exists on disk the file is removed and crio is reloaded.
+func (p *PinnedImageSetManager) ensureCrioPinnedImagesConfigFile(images []mcfgv1alpha1.PinnedImageRef) error {
+	hasCrioConfigFile, err := p.hasPinnedImagesConfigFile()
+	if err != nil {
+		return fmt.Errorf("failed to check crio config file: %w", err)
+	}
+
+	// if there are no images and the config file exists, remove it
+	if len(images) == 0 {
+		if hasCrioConfigFile {
+			// remove the crio config file and reload crio
+			if err := p.deleteCrioConfigFile(); err != nil {
+				return fmt.Errorf("failed to remove crio config file: %w", err)
+			}
+			return p.crioReload()
+		}
+		return nil
+	}
+
+	// ensure list elements are unique
+	// assume the number of duplicates is small so preallocate makes sense
+	// for larger lists
+	seen := make(map[string]struct{}, len(images))
+	unique := make([]string, 0, len(images))
+	for _, image := range images {
+		if _, ok := seen[image.Name]; !ok {
+			seen[image.Name] = struct{}{}
+			unique = append(unique, image.Name)
+		}
+	}
+
+	// create ignition file for crio drop-in config. this is done to use the existing
+	// atomic writer functionality.
+	ignFile, err := createCrioConfigIgnitionFile(unique)
+	if err != nil {
+		return fmt.Errorf("failed to create crio config ignition file: %w", err)
+	}
+
+	skipCertificateWrite := true
+	if err := writeFiles([]ign3types.File{ignFile}, skipCertificateWrite); err != nil {
```

Comment:
Theoretically can multiple routines be trying to write this file?

Reply:
in theory yes but the way it is written no. we only allow one workload at time. each sync will cancel the existing workload.